modules/oeFax/controller/AppDispatch.php

```diff
@@ -0,0 +1,248 @@
+    {
+        if ($param) {
+            return isset($this->_server[$param]) ?
+                $this->_server[$param] : $default;
+        }
+
+        return $this->_server;
+    }
+
+    public function getSession($param = null, $default = null)
+    {
+        if ($param) {
+            return isset($this->_session[$param]) ?
+                $this->_session[$param] : $default;
+        }
+
+        return $this->_session;
+    }
+
+    public function getCookie($param = null, $default = null)
+    {
+        if ($param) {
+            return isset($this->_cookies[$param]) ?
+                $this->_cookies[$param] : $default;
+        }
+
+        return $this->_cookies;
+    }
+
+    public function setHeader($params)
+    {
+        if (!headers_sent()) {
+            if (is_scalar($params)) {
+                header($params);
+            } else {
+                foreach ($params as $key => $value) {
+                    header(sprintf('%s: %s', $key, $value));
+                }
+            }
+        }
+
+        return $this;
+    }
+
+    public function setSession($key, $value)
+    {
+        $_SESSION[$key] = $value;
+        return $this;
+    }
+
+    public function setCookie($key, $value, $seconds = 3600)
+    {
+        $this->_cookies[$key] = $value;
+        if (!headers_sent()) {
+            setcookie($key, $value, time() + $seconds);
+            return $this;
+        }
+        return 0;
+    }
+
+    protected function saveSetup()
+    {
+        $username = $this->getRequest('username');
+        $ext = $this->getRequest('extension');
+        $password = $this->getRequest('password');
+        $appkey = $this->getRequest('key');
+        $appsecret = $this->getRequest('secret');
+        $production = $this->getRequest('production');
+        $smsNumber = $this->getRequest('smsnumber');
+        $smsMessage = $this->getRequest('smsmessage');
+        $smsHours = $this->getRequest('smshours');
+        $setup = array(
+            'username' => "$username",
+            'extension' => "$ext",
+            'password' => "$password",
+            'appKey' => "$appkey",
+            'appSecret' => "$appsecret",
+            'server' => !$production ? 'https://platform.devtest.ringcentral.com' : "https://platform.ringcentral.com",
+            'portal' => !$production ? "https://service.devtest.ringcentral.com/" : "https://service.ringcentral.com/",
+            'smsNumber' => "$smsNumber",
+            'production' => $production,
+            'redirect_url' => $this->getRequest('redirect_url'),
+            'smsHours' => $smsHours,
+            'smsMessage' => $smsMessage
+        );
+        $baseDir = $GLOBALS['OE_SITE_DIR'] . DIRECTORY_SEPARATOR . "messageStore";
+        $this->baseDir = $baseDir;
+
+        $cacheDir = $GLOBALS['OE_SITE_DIR'] . '/documents/logs_and_misc/_cache';
+        $fn = self::getServiceType() === '1' ? '/_credentials.php' : '/_credentials_twilio.php';
```

Comment:
TODO
for me to figure out why cache is needed for the setup after complete review

Reply:
It houses the encrypted credentials and authentication tokens.

Reply:
I can put it back in messageStore where I had it till recent

Reply:
Regarding messageStore:
All writable folders need to be within the documents directory (that is the only place where openemr has permission to write).
Regarding _cache directory, to avoid issues, recommend making that directory and placing a README file in it with quick sentence on purpose (this _cache directory will be a nice place for things like you are doing with the fax server stuff).

Reply:
Also, forgot to mention, for directory names (ie. messageStore), avoid caps (unless for class stuff).